resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -829,50 +967,176 @@ void add(DnsRecord r) {
+                servers++;
+            }
+        }
+
+        void handleWithAdditional(
+                DnsNameResolver parent, DnsRecord r, AuthoritativeDnsServerCache authoritativeCache) {
+            // Just walk the linked-list and mark the entry as handled when matched.
+            AuthoritativeNameServer serverName = head;
+
+            String nsName = r.name();
+            InetAddress resolved = decodeAddress(r, nsName, parent.isDecodeIdn());
+            if (resolved == null) {
+                // Could not parse the address, just ignore.
+                return;
+            }
+
+            while (serverName != null) {
+                if (serverName.nsName.equalsIgnoreCase(nsName)) {
+                    if (serverName.address != null) {
+                        // We received multiple ADDITIONAL records for the same name.
+                        // Search for the last we insert before and then append a new one.
+                        while (serverName.next != null && serverName.next.isCopy) {
+                            serverName = serverName.next;
+                        }
+                        AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                        server.next = serverName.next;
+                        serverName.next = server;
+                        serverName = server;
+
+                        servers++;
+                    }
+                    // We should replace the TTL if needed with the one of the ADDITIONAL record so we use
+                    // the smallest for caching.
+                    serverName.update(parent.newRedirectServerAddress(resolved), r.timeToLive());
+
+                    // Cache the server now.
+                    cache(serverName, authoritativeCache, parent.executor());
+                    return;
+                }
+                serverName = serverName.next;
             }
         }
 
-        // Just walk the linked-list and mark the entry as removed when matched, so next lookup will need to process
-        // one node less.
-        AuthoritativeNameServer remove(String nsName) {
+        // Now handle all AuthoritativeNameServer for which we had no ADDITIONAL record
+        void handleWithoutAdditionals(
+                DnsNameResolver parent, DnsCache cache, AuthoritativeDnsServerCache authoritativeCache) {
             AuthoritativeNameServer serverName = head;
 
             while (serverName != null) {
-                if (!serverName.removed && serverName.nsName.equalsIgnoreCase(nsName)) {
-                    serverName.removed = true;
-                    return serverName;
+                if (serverName.address == null) {
+                    // These will be resolved on the fly if needed.
+                    cacheUnresolved(serverName, authoritativeCache, parent.executor());
+                    servers++;
```

Comment:
the `AuthoritativeNameServer ` isn't being modified here ... remove this increment?